platforms/core-execution/execution/README.md

```diff
@@ -38,96 +46,160 @@ Examples of units of work include:
+- **Up-to-Date Checks**: If the _execution state_ of the work is **up-to-date**, the execution is skipped.
+- **Build Cache**: If the output is not up-to-date locally, the engine searches for stored results in the _build cache,_ checking the local cache first and then the remote cache if necessary.
+- **Execution**: If no result is found in the build cache, the work is executed.
+
+### Origin Metadata
 
-_TBD_
+Each produced result carries with it some **origin metadata**. This is a map of key-value pairs that describe which build created the result, how long it took, and the cache key associated with the result.
+When a result is produced by the execution engine the origin metadata can be consulted to understand if the result was produced during this build, or if it was reused from a previous one.
+
+### Up-to-Date Checks (aka Incremental Build)
+
+> [!NOTE]
+> Incremental _build_ and incremental _work_ refer to different kinds of incrementality.
+> We should probably stop using incremental build as a term.
+
+This way we can compare the state of a unit of work before executing it to its state after its previous execution.
+If there are no changes to either its inputs or outputs, executing the work is not needed, and is thus skipped.
+
+* Input value changes are detected by comparing the `ValueSnapshot`s input value properties.
+
+* Input file changes are detected by comparing fingerprints of the input file properties.
+
+* Output changes are detected by comparing the snapshots of the output properties.
 
 ### The Build Cache
 
-_TBD_
+If the execution state of the work is out-of-date, we try to load outputs from the build cache.
+If we are allowed to use the cache, then first the local cache is consulted.
+If there is no entry with the cache key in the local cache, then we ask the remote cache.
+
+If there is a cache hit, we delete and unpack results to the output location.
+Local state is also cleared up when a result is loaded.
+
+If there's no cache hit, or we were not allowed to load from cache, we continue with execution.
+Once execution finished, we check if we are allowed to store in the local cache.
+If we are allowed, we store in the local and remote caches.
+
+> [!NOTE]
+> For tasks we have an internal mechanism that can prevent storing the results of an executed task in the build cache.
+> This internal feature was added and is only used by an optimization in Develocity's predictive test selection.
 
 ### Execution Modes
 
-Note that this differs from _incremental build,_ which involves skipping the entire execution of an up-to-date unit of work.
+When it comes to actually execution work, the execution engine supports two kinds of work: _mutable_ and _immutable._ _Mutable_ work can be executed _incrementally_ or _non-incrementally_ based on which of its inputs have changed compared to a previous execution.
+
+- **Immutable Work** All inputs are **identity inputs**; the complete input set defines the work.
```

Comment:
This does not seem to be true about the Dependency Accessors unit of work, despite it being `ImmutableUnitOfWork`

https://github.com/gradle/gradle/blob/80a62ea50c7845a2a3ebbbb8c8523a05d4f5e67c/platforms/software/dependency-management/src/main/java/org/gradle/api/internal/catalog/DefaultDependenciesAccessors.java#L424-L442